### PR TITLE
refactor: don't convert schema to protobuf just for pre-order traversal of fields

### PR DESCRIPTION
**Pull request by westonpace**

### CODE REVIEW

**Comment by westonpace** [comment]
There was a TODO for this

File: rust/lance/src/datatypes/schema.rs [hunk 1]
```diff
@@ -96,15 +150,27 @@ impl Schema {
...
+    /// Returns a new schema that only contains the fields in `column_ids`.
+    ///
+    /// This projection can filter out both top-level and nested fields
+    pub fn project_by_ids(&self, column_ids: &[i32]) -> Self {
+        let filtered_fields = self
+            .fields
             .iter()
-            .filter(|p| column_ids.contains(&p.id))
-            .cloned()
+            .filter_map(|f| f.project_by_filter(&|f| column_ids.contains(&f.id)))
```
Comment (eddyxu) [inline rust/lance/src/datatypes/schema.rs:168]: i forget. does it return the path from the root node to deep nested field if only the leave fields are specified via `column_ids` ?

File: rust/lance/src/datatypes/schema.rs [hunk 2]
```diff
@@ -172,11 +238,9 @@ impl Schema {
             })
     }
 
-    /// Recursively collect all the field IDs,
+    /// Recursively collect all the field IDs, in pre-order traversal order.
     pub(crate) fn field_ids(&self) -> Vec<i32> {
-        // TODO: make a tree traversal iterator.
```
Comment (eddyxu) [inline rust/lance/src/datatypes/schema.rs:177]: Thanks for finally making this happen :)

File: rust/lance/src/datatypes/schema.rs (hunk 1 above)
Comment (westonpace) [reply rust/lance/src/datatypes/schema.rs:168]: Good catch.  I added a test case for this and fixed it.  In order to maintain the old behavior we need to keep parent fields if they have any matching children.  I think this is the most intuitive behavior as well.  We can always add support for collapsing a schema (e.g. remove structs with only one child) later as a different method.

File: rust/lance/src/datatypes/schema.rs [hunk 3]
```diff
@@ -35,6 +35,60 @@ pub struct Schema {
     pub metadata: HashMap<String, String>,
 }
 
+/// State for a pre-order DFS iterator over the fields of a schema.
+struct SchemaFieldIterPreOrder<'a> {
+    schema: &'a Schema,
+    field_stack: Vec<&'a Field>,
+    idx_stack: Vec<usize>,
+    top_level_idx: usize,
```
Comment (wjones127) [inline rust/lance/src/datatypes/schema.rs:43]: bikeshed: Would be simpler just to have 1 stack: insert fields in reverse order, pop from the back, if nested push children back in reverse order?

**Review by wjones127 — APPROVED** [review]

File: rust/lance/src/datatypes/schema.rs (hunk 3 above)
Comment (westonpace) [reply rust/lance/src/datatypes/schema.rs]: That was a lot simpler.  Great idea!

**Review by wjones127 — APPROVED** [review]

**Pull request merged**